Chap_API_Struct.tex

```diff
@@ -969,6 +969,11 @@ \subsection{Value Structure}
 \} pmix_value_t;
 }
 
+The size of the \refstruct{pmix_value_t} must remain constant on a given platform regardless of the compiler.
+The \refstruct{pmix_value_t} is \code{32} bytes with \code{24} bytes reserved for the \code{data} union member.
+An implementation may add a padding member to the structure to achieve the required 32-byte size.
```

Comment:
Are we saying it should be 32 bytes regardless of platform or just on x86-64 (or ppc64le 😄 )? FWIW, if I build a 32-bit x86 executable with the current ABI headers,`pmix_value_t` is 16 bytes and the data portion is 12.

Reply:
Good point - we definitely don't want to force 32-bit builds to waste that space.

Reply:
Good point. Do you all have a suggestion on a re-wording?

Reply:
Actually, let's take a step back. It occurs to me that we may be over complicating things here. Let's take a look at what we mean by "ABI".

I think we all agree that we should take the "build" meaning for ABI, so let's take that as a starting point. This means that a program is assured that they can access a library of the same ABI they built against, with some assurance that symbols will resolve, constants will match in value, and structs will have the same size/ordering of fields.

There is nothing in that statement to say a 64-bit program could use a 32-bit library, or vice versa. Likewise, there is no statement that indicates you can move your app across architecture without change - which is good because that is pretty much impossible anyway. If you move from x86-64 to ARM, you are going to have to recompile.

So any statements about padding should be restricted by those constraints. It should be adequate to state that the padding should align values with the native (64 or 32-bit) addressing scheme.

However, there is also nothing that proposes that a client app can interact with a server based on a different implementation. In the case of a shared memory subsystem, that would be highly unlikely to work anyway as the details of internal implementation are probably different. For non-shared memory interactions, there is no standard wire protocol - and even if there were, it wouldn't affect the ABI as both sides instantiate the objects passed over the wire themselves.

So do we even need padding? I'm thinking we don't as the only reason to pad was the shared memory subsystem, and that actually is an implementation-internal subsystem - and not an ABI. In other words, the ABI should not apply to cross-implementation interactions - it should be restricted to the interface between a program and an implementation's library.

Reply:
The primary use case that we have for the ABI is to be able to build against the PMIx headers, then link with various PMIx implementations. This would allow a tool to build without linking to a specific PMIx implementation then use `dlopen/dlsym` to pick up the library symbols at runtime.

On a given architecture+compiler the sizeof the `pmix_value_t` should not change if the structure stays the same as it is defined in the standard. If additional fields are added to the union by an implementation beyond what is defined in the standard then it should not exceed the sizeof the data union as it exists in the standard.

Maybe that last sentence is all that we need? That will ensure that a tool built against ABI vX.Y will work with an implementation that supports ABI vX.Y even if that implementation has extra elements in the data union - since that union has not grown. So any arrays of structures that the tool has created statically are still at the correct size/offset.

I'd like to discuss this today in the call to reach a resolution.

Reply:
> Maybe that last sentence is all that we need? That will ensure that a tool built against ABI vX.Y will work with an implementation that supports ABI vX.Y even if that implementation has extra elements in the data union - since that union has not grown. So any arrays of structures that the tool has created statically are still at the correct size/offset.

IMO, yes. That should be all we need. As long as the size (and offset) of the elements in the structure are consistent, then no pad should be needed.

Reply:
Note from April 14, 2022 meeting:
> The PMIx ABI is platform-specific. The headers are defined, but the realization of the ABI in binary form is platform-specific. Therefore care must be taken with data structures to not change the position of members, size of the structure or members in the structure, and the offset of the members in the structure for that platform.

Reply:
I just pushed this change as a new commit.